### PR TITLE
Ltca

### PR DESCRIPTION
**Pull request by frangipane**

### CODE REVIEW

File: content/articles/methods/ltca.md [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+
+[TOC]
+
+# Summary
+
+Standard reinforcement learning algorithms struggle with poor sample efficiency in the presence of sparse rewards with long temporal delays between action and effect. To address the long term credit assignment problem, we build on the work of [1] to use “temporal reward transport” (TRT) to augment the immediate rewards of significant state-action pairs with rewards from the distant future using an attention mechanism to identify candidates for TRT. A series of gridworld experiments show clear improvements in learning when TRT is used in conjunction with a standard advantage actor critic algorithm.
+
+# Introduction
+
+Episodic reinforcement learning models the interaction of an agent with an environment as a Markov Decision Process with a finite number of time steps $T$.  The agent transitions from one state to another based on its actions and the environment dynamics $p(s’,r|s, a)$, where the transitions to the next state $s'$ and the reward picked up along the way, $r$, are modeled as a joint distribution over the previous state $s$ and action $a$.  In general, the agent does not have access to an exact model of the environment.
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: what is p? a probability? the "policy"?

File: content/articles/methods/ltca.md [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+
+The agent's goal is to maximize its cumulative rewards, the discounted returns $G_t$,
+
+\begin{eqnarray}\label{return} \tag{1}
+G_t := R_{t+1} + \gamma R_{t+2} + … = \sum_{k=0}^T \gamma^k R_{t+k+1}
+\end{eqnarray}
+
+where $0 \leq \gamma \leq 1$, and $R_{t}$ is the reward at time $t$.  In episodic RL, the discount factor $\gamma$ is often used to account for uncertainty in the future, to favor rewards now vs. later, and as a variance reduction technique, e.g. in policy gradient methods [2, 3].
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md:25]: reinforcement learning (RL)

File: content/articles/methods/ltca.md [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+
+Using a discount factor $\gamma < 1$ introduces a timescale by exponentially suppressing rewards in the future by $\exp(-n/\tau_{\gamma})$.  The number of timesteps it takes for a reward to decay by $1/e$ is $\tau_{\gamma} = 1/(1-\gamma)$, which follows from
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: not sure if tau is normal - usually tau has units of time.  This seems to be unitless, why not just n_\gamma?

File: content/articles/methods/ltca.md [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+
+\begin{align}\label{discount-timescale}\tag{2}
+\gamma ^ n \approx \frac{1}{e} = \lim_{n \rightarrow \infty} \left(1 - \frac{1}{n} \right)^n
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md:30]: where is tau here?

File: content/articles/methods/ltca.md [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+\end{align}
+
+* * * * *
+
+The state value function $v_{\pi}(s)$ is the expected return when starting in state $s$, following policy $\pi(a|s) := p(a|s)$, a deterministic or stochastic function of the current state.
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: a deterministic or stochastic function = a function ?

File: content/articles/methods/ltca.md [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+
+\begin{eqnarray}\label{state-value} \tag{3}
+v_{\pi}(s) = \mathbb{E}_{\pi}[G_t | S_t = s]
+\end{eqnarray}
+
+Policy gradient algorithms improve the policy by using gradient ascent along the gradient of the value function.
+
+\begin{eqnarray}\label{policy-gradient} \tag{4} 
+\nabla_{\theta} v_0 = \mathbb{E}_{\tau \sim \pi}\left[\sum_{t=0}^T \nabla_{\theta} \log \pi_{\theta}(a_t | s_t) \mathcal{R}(\tau)\right],
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: v_0 or v(0)?

or is this some sort of initial

I dunno what tau \sim pi means either...

File: content/articles/methods/ltca.md [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+\end{eqnarray}
+
+where $\tau$ describes the agent's trajectory, beginning from state $s_0$, and $\mathcal{R}(\tau)$ is a function of the rewards obtained along the trajectory.  In practice, policy gradients approximate the expected value in (\ref{policy-gradient}) by sampling, which results in very high variance estimates of the gradient.
+
+Common techniques to reduce the variance of the estimated policy gradient include
+
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md:50]: citations for each of these or are they just well-known (cite a text book on the whole thing?)

File: content/articles/methods/ltca.md [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+1. only assigning credit for future rewards, the "rewards-to-go", to a particular action of rewards for the entire trajectory, which include rewards obtained before that action
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: I don't get it.

File: content/articles/methods/ltca.md [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+2. subtracting a baseline from the rewards weight that is independent of action.  Oftentimes, this baseline is the value function in (\ref{state-value}), which also has to be estimated, giving rise to actor-critic algorithms that learn a policy and value function in parallel.
+3. using a large batch size.
+4. using the value function (\ref{state-value}) to bootstrap the returns some number of steps into the future instead of using the full raw discounted return.  For example, one-step bootstrapping would approximate the discounted returns in (\ref{return}) as
+
+\begin{eqnarray}\label{bootstrap} \tag{5}
+G_t = R_{t+1} + \gamma R_{t+2} +  \gamma^2 R_{t+3} + \ldots \approx R_{t+1} + \gamma V(t+1).
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: what's V(t+1)?  Is it supposed to be v like in \ref{state-value} ??

File: content/articles/methods/ltca.md [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+\end{eqnarray}
+
+All of these techniques typically make use of discounting, so an action receives little credit for rewards that happen after some multiple of $\tau_{\gamma}$ timesteps in the future, making it challenging for standard reinforcement learning algorithms to learn effective policies in situations where action and effect are separated by long temporal delays, resulting in poor sample efficiency.
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: get rid of "some multiple of" for brevity

I also don't understand this clause " resulting in poor sample efficiency" and what it brings to the sentence.

File: content/articles/methods/ltca.md [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+
+Concretely, the gridworld environment consists of:
+
+(1) Empty grid with key: agent can pick up the key but receives no immediate reward for picking it up.
+
+<p align="center">
+<img src="images/ltca_gridworld_p1.png" alt="Phase 1" style="width:190px;">
+</p>
+
+(2) Distractor phase: Agent engages with distractors (toggling open gifts) that yield immediate rewards.
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: "toggling open gifts?"  

gifts that can be opened, yielding immediate rewards.

File: content/articles/methods/ltca.md [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+<img src="images/ltca_gridworld_p2.png" alt="Phase 1" style="width:190px;">
+</p>
+
+(3) Delayed reward phase: Agent should move to a green goal grid cell as quickly as possible.  If the agent is carrying the key when it reaches the goal, it is rewarded extra points.
+
+<p align="center">
+<img src="images/ltca_gridworld_p3.png" alt="Phase 1" style="width:190px;">
+</p>
+
+The agent remains in each phase for a fixed period of time, regardless of how quickly it finishes the intermediate task, and then teleports to the next phase.  The environment resets with a different random seed that randomizes the placement of the key in phase 1 and distractor objects in phase 2.
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: "The agent remains in each phase for a fixed period of time, regardless of how quickly it finishes the intermediate task, and then teleports to the next phase"

they why should it move to the green goal grid cell "as quickly as possible"?

File: content/articles/methods/ltca.md [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+
+Note, unlike the Reconstructive Memory Agent in [1], our agent does not have the ability to encode and reconstruct memories, and our environment is not set up to test for that ability.
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: Is this particularly relevant?  maybe it should be a footnote?

File: content/articles/methods/ltca.md [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+
+## Experiments
+
+In all experiments, the time spent in phase 1 and phase 3 is fixed, as well as the distal reward in phase 3: 5 points for reaching the goal without a key and 20 points for reaching the goal carrying a key (with a small penalty proportional to step count to encourage moving quickly to the goal).
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: how many points per gift in phase 2?

File: content/articles/methods/ltca.md [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+
+Each experiment varies a particular parameter in the second phase to evaluate the effect on learning for the baseline A2C algorithm compared to A2C enhanced with TRT.
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: which parameter?

Comment (dustinmcintosh) [reply content/articles/methods/ltca.md]: I guess time delay, based on next section, but maybe just say that here?

File: content/articles/methods/ltca.md [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+
+The focus our evaluation for each experiment is on the last reward (the distal reward) obtained in phase 3, although we also show the overall returns to make sure the agent is also learning a policy that also maximizes the rewards in phase 2.
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: get rid of first "also"

File: content/articles/methods/ltca.md [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+<img src="images/ltca_reward_expt_plots.png" alt="P2 reward size expt" style="width:800px;">
+</p>
+
+<small>Fig 2. Returns in phase 3 for rewards in phase 2 of size 0, 5, and 8.
+</small>
+
+
+### Distractor reward variance
+
+We fix the mean reward size of the gifts in phase 2 at 5, but the variance of the rewards ranges from 0, to 1.33, to 8.33, corresponding to drawing each reward from a uniform distribution centered at 5, with minimum and maximum ranges of [5, 5], [3, 7], and [0, 10], respectively.
```
Comment (dustinmcintosh) [inline content/articles/methods/ltca.md]: I'm not sure what we get out of the variances ... the ranges are more useful.

Comment (frangipane) [reply content/articles/methods/ltca.md]: reworded

File: content/articles/methods/ltca.md (hunk 12 above)
Comment (frangipane) [reply content/articles/methods/ltca.md]: Good point, removed "as quickly as possible" here to be less confusing, but this is explained in more detail in the methods section.

File: content/articles/methods/ltca.md (hunk 8 above)
Comment (frangipane) [reply content/articles/methods/ltca.md]: reworded

File: content/articles/methods/ltca.md (hunk 9 above)
Comment (frangipane) [reply content/articles/methods/ltca.md]: clarified it's the estimate (since we can't actually calculate the expected value, only a sampled approx)

**Pull request merged**